php/Blockonomics.php

```diff
@@ -238,4 +294,334 @@ public function testSetup()
+        $order['timestamp'] = time();
+        return $order;
+    }
+
+    // Get any existing unpaid order for the crypto
+    public function get_waiting_order($orders, $order_id, $crypto){
+        foreach ($orders as $address => $order) {
+            if($order['crypto'] && $crypto) {
+                // Check if order and has expired
+                if ( $order['timestamp'] <= time() - get_option("blockonomics_timeperiod") * 60 ) {
+                    $order = $this->update_waiting_order($order);
+                }
+                return $order;
+            }
+        }
+        return false;
+    }
+
+    // Save the new address to the WooCommerce order
+    public function record_address($order_id, $crypto, $address){
+        update_post_meta($order_id, $crypto .'_address', $address);
+    }
+
+    public function create_new_order($order_id, $crypto){
+        $responseObj = $this->new_address(get_option("blockonomics_callback_secret"), $crypto);
+        if($responseObj->response_code != 200) {
+            exit();
+        }
+        $address = $responseObj->address;
+
+        $order = array(
+                'order_id'           => $order_id,
+                'status'             => -1,
+                'crypto'             => $crypto,
+                'address'            => $address
+        );
+        $order = $this->update_waiting_order($order);
+
+        $this->record_address($order_id, $crypto, $address);
+        return $order;
+    }
+
+    // Load the the checkout template in the page
+    public function load_checkout_template($order_id, $crypto){
+        // Check to send the user to nojs page
+        if($this->is_nojs_active()){
+            // Create or update the order for the nojs template
+            $this->process_order($order_id, $crypto);
+            $this->load_blockonomics_template('nojs_checkout');
+        }else{
+            $this->load_blockonomics_template('checkout');
+        }
+    }
+
+    // Load the the crypto options template in the page
+    public function load_crypto_options_template(){
+        // Check to send the user to nojs page
+        if($this->is_nojs_active()){
+            $this->load_blockonomics_template('nojs_crypto_options');
+        }
+        else{
+            $this->load_blockonomics_template('crypto_options');
+        }
+    }
+
+    // Redirect the user to the woocommerce finish order page
+    public function redirect_finish_order($order_id){
+        $wc_order = new WC_Order($order_id);
+        wp_redirect($wc_order->get_checkout_order_received_url());
+        exit();
+    }
+
+    // Fetch all the crypto orders linked to the order id
+    public function get_all_orders_by_id($orders, $order_id){
+        if(isset($orders[$order_id])){
+            return $orders[$order_id];
+        }
+        return false;
+    }
+
+    // Check and update the crypto order or create a new order
+    public function process_order($order_id, $crypto){
```

Comment:
Though the logic is right, it is confusing and a bit complex. Here is a simple logic (it doesn't cover some minor cases, but should be ok):
$order = get order with *order_id* and *crypto*
if ($order)
 $order = calculate_order_params($order)
else
  $order = create_new_order(order_id, crypto)

Notes:
*calculate_order_params* will not calculate if order status!=-1 or timestamp has not expired
For payment in process there is no extra logic required. When client gets these params along with correct timestamp, it will automatically redirect to order confirmation page after receiving tx from blockonomics

Reply:
Updated in 69662b2